File: 6/benchmark.rb
require 'benchmark'
require './potapkin'
require './guzhin'

def expect_to_equal(expected, actual, label = nil)
  if expected != actual
    puts "#{label || ""} FAILURE!!! [#{actual}] is not equal to [#{expected}]" 
  end
end

def tests(mod)
  fn = mod.method(:roman)
  expect_to_equal("I", fn.call(1), "1 = I");
  expect_to_equal("II", fn.call(2), "2 = II");
  expect_to_equal("III", fn.call(3), "3 = III");
  expect_to_equal("IV", fn.call(4), "4 = IV");
  expect_to_equal("V", fn.call(5), "5 = V");
  expect_to_equal("VI", fn.call(6), "6 = VI");
  expect_to_equal("VII", fn.call(7), "7 = VII");
  expect_to_equal("VIII", fn.call(8), "8 = VIII");
  expect_to_equal("IX", fn.call(9), "9 = IX");
  expect_to_equal("X", fn.call(10), "10 = X");
  expect_to_equal("XI", fn.call(11), "11 = XI");
  expect_to_equal("XII", fn.call(12), "12 = XII");
  expect_to_equal("XIII", fn.call(13), "13 = XIII");
  expect_to_equal("XIV", fn.call(14), "14 = XIV");
  expect_to_equal("XV", fn.call(15), "15 = XV");
  expect_to_equal("XVI", fn.call(16), "16 = XVI");
  expect_to_equal("XVII", fn.call(17), "17 = XVII");
  expect_to_equal("XVIII", fn.call(18), "18 = XVIII");
  expect_to_equal("XIX", fn.call(19), "19 = XIX");
  expect_to_equal("XX", fn.call(20), "20 = XX");
  expect_to_equal("XXI", fn.call(21), "21 = XXI");
  expect_to_equal("XXXII", fn.call(32), "32 = XXXII");
  expect_to_equal("XLIV", fn.call(44), "44 = XLIV");
  expect_to_equal("LV", fn.call(55), "55 = LV");
  expect_to_equal("LXVI", fn.call(66), "66 = LXVI");
  expect_to_equal("XCIX", fn.call(99), "99 = XCIX");
  expect_to_equal("C", fn.call(100), "100 = C");
  expect_to_equal("CI", fn.call(101), "101 = CI");
  expect_to_equal("CCII", fn.call(202), "202 = CCII");
  expect_to_equal("CCCXLIV", fn.call(344), "344 = CCXLIV");
  expect_to_equal("CDXLIV", fn.call(444), "444 = CDXLIV");
  expect_to_equal("D", fn.call(500), "500 = D");
  expect_to_equal("DCXLIX", fn.call(649), "649 = DCXLIX");
  expect_to_equal("CM", fn.call(900), "900 = CM");
  expect_to_equal("MMMCMXCIX", fn.call(3999), "3999 = MMMCMXCIX");
end

def measure(label, callable, divide_by = 1)
  time = Benchmark.measure do
    callable.call();
  end
  time.real.to_f / divide_by
end

def case1(mod)
  fn = mod.method(:roman)

  measure(
    "Find the roman interpretation of 10_000_000 of 3999",
    lambda {
      10_000_000.times do
        throw StandardError.new("CASE I failed") if "MMMCMXCIX" != fn.call(3999)
      end
    },
    10_000_000,
  )
end

def case2(mod)
  fn = mod.method(:roman)

  measure(
    "Find the roman interpretation of 10_000_000 of 1",
    lambda {
      10_000_000.times do
        throw StandardError.new("CASE II failed") if "I" != fn.call(1)
      end
    },
    10_000_000,
  )
end

SOLUTIONS = [
  {
    author: "Kirill Potapkin",
    module: Potapkin,
  },
  {
    author: "Serhei Guzhin",
    module: Guzhin,
  },
];

def test_solution(solution)
  tests(solution[:module])
  case1_ms = case1(solution[:module])
  case2_ms = case2(solution[:module])
  puts [
    "#{solution[:author]} (Ruby)",
    (case1_ms * 1_000_000_000).round(2),
    (case2_ms * 1_000_000_000).round(2)
  ].join(';').gsub('.', ',')
end

SOLUTIONS.each do |solution|
  test_solution(solution);
end

File: 6/benchmark.js
import { expectToEqual, measure, testSolution } from "../framework/utils.js";
import * as malyshev from "./malyshev.js";
import * as malyshev2 from "./malyshev2.js";
import * as kazantsev from "./kazantsev.js";
import * as larkin from "./larkin.js";
import * as yakuncheva from "./yakuncheva.js";
import * as kazantsev2 from "./kazantsev2.js";
import * as zakharov from "./zakharov.js";
import * as kazantsev3 from "./kazantsev3.js";

const tests = (mod) => {
  expectToEqual("I", mod.roman(1), "1 = I");
  expectToEqual("II", mod.roman(2), "2 = II");
  expectToEqual("III", mod.roman(3), "3 = III");
  expectToEqual("IV", mod.roman(4), "4 = IV");
  expectToEqual("V", mod.roman(5), "5 = V");
  expectToEqual("VI", mod.roman(6), "6 = VI");
  expectToEqual("VII", mod.roman(7), "7 = VII");
  expectToEqual("VIII", mod.roman(8), "8 = VIII");
  expectToEqual("IX", mod.roman(9), "9 = IX");
  expectToEqual("X", mod.roman(10), "10 = X");
  expectToEqual("XI", mod.roman(11), "11 = XI");
  expectToEqual("XII", mod.roman(12), "12 = XII");
  expectToEqual("XIII", mod.roman(13), "13 = XIII");
  expectToEqual("XIV", mod.roman(14), "14 = XIV");
  expectToEqual("XV", mod.roman(15), "15 = XV");
  expectToEqual("XVI", mod.roman(16), "16 = XVI");
  expectToEqual("XVII", mod.roman(17), "17 = XVII");
  expectToEqual("XVIII", mod.roman(18), "18 = XVIII");
  expectToEqual("XIX", mod.roman(19), "19 = XIX");
  expectToEqual("XX", mod.roman(20), "20 = XX");
  expectToEqual("XXI", mod.roman(21), "21 = XXI");
  expectToEqual("XXXII", mod.roman(32), "32 = XXXII");
  expectToEqual("XLIV", mod.roman(44), "44 = XLIV");
  expectToEqual("LV", mod.roman(55), "55 = LV");
  expectToEqual("LXVI", mod.roman(66), "66 = LXVI");
  expectToEqual("XCIX", mod.roman(99), "99 = XCIX");
  expectToEqual("C", mod.roman(100), "100 = C");
  expectToEqual("CI", mod.roman(101), "101 = CI");
  expectToEqual("CCII", mod.roman(202), "202 = CCII");
  expectToEqual("CCCXLIV", mod.roman(344), "344 = CCXLIV");
  expectToEqual("CDXLIV", mod.roman(444), "444 = CDXLIV");
  expectToEqual("D", mod.roman(500), "500 = D");
  expectToEqual("DCXLIX", mod.roman(649), "649 = DCXLIX");
  expectToEqual("CM", mod.roman(900), "900 = CM");
  expectToEqual("MMMCMXCIX", mod.roman(3999), "3999 = MMMCMXCIX");
};

const case1 = (mod) => {
  return measure(
    "Find the roman interpretation of 10_000_000 of 3999",
    () => {
      for (let i = 0; i < 10_000_000; i++) {
        if ("MMMCMXCIX" !== mod.roman(3999)) {
          throw new Error("CASE I failed");
        }
      }
    },
    10,
    10_000_000
  );
};

const case2 = (mod) => {
  return measure(
    "Find the roman interpretation of 10_000_000 of 1",
    () => {
      for (let i = 0; i < 10_000_000; i++) {
        if ("I" !== mod.roman(1)) {
          throw new Error("CASE II failed");
        }
      }
    },
    10,
    10_000_000
  );
};
const solutions = [
  {
    author: "Andrei Malyshev",
    module: malyshev,
  },
  {
    author: "Andrei Malyshev (smartass cheat)",
    module: malyshev2,
  },
  {
    author: "Vlad Kazantsev",
    module: kazantsev,
  },
  {
    author: "Vyacheslav Larkin",
    module: larkin,
  },
  {
    author: "Irina Yakuncheva",
    module: yakuncheva,
  },
  {
    author: "Vlad Kazantsev 2",
    module: kazantsev2,
  },
  {
    author: "Mikhail Zakharov",
    module: zakharov,
  },
  {
    author: "Vlad Kazantsev 3",
    module: kazantsev3,
  },
];

solutions.forEach((solution) => {
  testSolution(solution, tests, [
    { case: case1, label: "Worst case" },
    { case: case2, label: "Best case" },
  ]);
});
